server/daos/booking_dao.js

```diff
@@ -40,27 +40,61 @@ exports.clearBookingTable = function () {
 
 
 //it allows you to insert a new booking
-exports.insertBooking = function ({ lecture_id, student_id }) {
+exports.insertBooking = async function ({ lecture_id, student_id }) {
     return new Promise((resolve, reject) => {
-        const sql = 'INSERT INTO Bookings(lecture_id,student_id) VALUES(?,?)'
-        db.run(sql, [lecture_id, student_id], function (err) {
-            if (err) {
-                const sql2 = 'UPDATE Bookings SET deleted_at= NULL WHERE lecture_id= ? AND student_id= ? AND deleted_at IS NOT NULL'
-                db.run(sql2, [ lecture_id, student_id], function (err2) {
-                    if (err2 || this.changes===0) {
-                        reject(err2);
+
+        // First check if the student has already booked for this lecture
+        let alreadyBookedQyery  =  `
+        SELECT *
+        FROM Bookings B
+        WHERE lecture_id = ? AND student_id = ? AND deleted_at IS NULL
+    `;
+        db.get(alreadyBookedQyery,[lecture_id,student_id],function(err,row){
+            if(row || err){
+                reject(Error('Already booked'));
+            }else{
+                const sql = `
+                SELECT COUNT(*) >= R.seats is not null and COUNT(*) >= R.seats as waiting
+                FROM Bookings B, Lectures L, Rooms R
+                WHERE B.lecture_id = L.id AND L.room_id = R.id AND B.lecture_id = ?1 
+                    AND B.deleted_at IS NULL AND B.waiting = 0
+            `;
+            db.get(sql, [lecture_id], function (err, row)  {
+                const { waiting } = row;
+                const sql2 = `
+                    INSERT INTO Bookings(lecture_id, student_id, waiting)
+                    VALUES(?1, ?2, ?3)
+                `;
+                db.run(sql2, [lecture_id, student_id, waiting], function (err2) {
+                    if (err2) {
+                        // TODO: check that updated_at is actually updated
+                        const sql3 = `
+                          UPDATE Bookings
+                          SET deleted_at = NULL, waiting = ?3
+                          WHERE lecture_id = ?1 AND student_id = ?2
+                        `;
+                        db.run(sql3, [ lecture_id, student_id, waiting], function (err3) {
+                            if (err3 || this.changes===0) {
+                                reject(err3);
+                            }
+                            else{
+                                const booking = createBooking(lecture_id, student_id, !!waiting);
```

Comment:
Why do we have !!waiting? is not the same of 'waiting'?

Reply:
it is probably used to force the type to boolean

But I haven't written that myself, the change probably comes from @bit97

Reply:
`waiting` can be (is?) an integer value, so that forces it to be a boolean (works on all nullish values)